packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -1010,12 +1014,20 @@ abstract class TestRunner {
     );
   }
 
-  private validateActiveTargets(): void {
+  private async validateActiveTargets(): Promise<void> {
     if (!this.isPrimaryClient) {
       expect(this.connection.activeTargets).to.be.empty;
       return;
     }
 
+    // In multi-tab mode, we cannot rely on the `waitForWatchOpen` call in
+    // `doUserListen` since primary tabs may execute queries from other tabs
+    // without any direct user interaction.
+    if (!obj.isEmpty(this.expectedActiveTargets)) {
+      await this.connection.waitForWatchOpen();
+      await this.queue.drain();
+    }
```

Comment:
Hrm.  It seems a bit awkward to do this as part of every step (and do we end up draining the queue an extra time?)...  But I guess we don't have a mechanism to only do it on client changes or primary changes or something?

I don't care too much, but maybe it's worth a `// TODO(multitab): We'd ideally only do this after a primary change.` or something?

Reply:
I looked at various places to put this... but there's no real good alternative. We could wait for the watch stream to open when we switch the clients, but it would seem really out of place (especially since we have to check that there are expected targets).

TODO added.